src/ray/gcs/gcs_client/service_based_accessor.h

```diff
@@ -0,0 +1,319 @@
+
+#include "src/ray/gcs/accessor.h"
+#include "src/ray/gcs/subscription_executor.h"
+
+namespace ray {
+namespace gcs {
+
+class ServiceBasedGcsClient;
+
+/// \class ServiceBasedJobInfoAccessor
+/// ServiceBasedJobInfoAccessor is an implementation of `JobInfoAccessor`
+/// that uses GCS Service as the backend.
+class ServiceBasedJobInfoAccessor : public JobInfoAccessor {
+ public:
+  explicit ServiceBasedJobInfoAccessor(ServiceBasedGcsClient *client_impl);
+
+  virtual ~ServiceBasedJobInfoAccessor() = default;
+
+  Status AsyncAdd(const std::shared_ptr<JobTableData> &data_ptr,
+                  const StatusCallback &callback) override;
+
+  Status AsyncMarkFinished(const JobID &job_id, const StatusCallback &callback) override;
+
+  Status AsyncSubscribeToFinishedJobs(
+      const SubscribeCallback<JobID, JobTableData> &subscribe,
+      const StatusCallback &done) override;
+
+ private:
+  ServiceBasedGcsClient *client_impl_{nullptr};
+
+  typedef SubscriptionExecutor<JobID, JobTableData, JobTable> JobSubscriptionExecutor;
+  JobSubscriptionExecutor job_sub_executor_;
+};
+
+/// \class ServiceBasedActorInfoAccessor
+/// ServiceBasedActorInfoAccessor is an implementation of `ActorInfoAccessor`
+/// that uses GCS Service as the backend.
+class ServiceBasedActorInfoAccessor : public ActorInfoAccessor {
+ public:
+  explicit ServiceBasedActorInfoAccessor(ServiceBasedGcsClient *client_impl);
+
+  virtual ~ServiceBasedActorInfoAccessor() = default;
+
+  Status AsyncGet(const ActorID &actor_id,
+                  const OptionalItemCallback<rpc::ActorTableData> &callback) override;
+
+  Status AsyncRegister(const std::shared_ptr<rpc::ActorTableData> &data_ptr,
+                       const StatusCallback &callback) override;
+
+  Status AsyncUpdate(const ActorID &actor_id,
+                     const std::shared_ptr<rpc::ActorTableData> &data_ptr,
+                     const StatusCallback &callback) override;
+
+  Status AsyncSubscribeAll(
+      const SubscribeCallback<ActorID, rpc::ActorTableData> &subscribe,
+      const StatusCallback &done) override;
+
+  Status AsyncSubscribe(const ActorID &actor_id,
+                        const SubscribeCallback<ActorID, rpc::ActorTableData> &subscribe,
+                        const StatusCallback &done) override;
+
+  Status AsyncUnsubscribe(const ActorID &actor_id, const StatusCallback &done) override;
+
+  Status AsyncAddCheckpoint(const std::shared_ptr<rpc::ActorCheckpointData> &data_ptr,
+                            const StatusCallback &callback) override;
+
+  Status AsyncGetCheckpoint(
+      const ActorCheckpointID &checkpoint_id,
+      const OptionalItemCallback<rpc::ActorCheckpointData> &callback) override;
+
+  Status AsyncGetCheckpointID(
+      const ActorID &actor_id,
+      const OptionalItemCallback<rpc::ActorCheckpointIdData> &callback) override;
+
+ private:
+  ServiceBasedGcsClient *client_impl_{nullptr};
+
+  ClientID subscribe_id_{ClientID::FromRandom()};
```

Comment:
Why not move this mutable initialization to constructor list.

Reply:
fixed